rfd/0163-vnet.md

```diff
@@ -0,0 +1,416 @@
+e.g. `api.teleport.example.com`.
+That is the URL you would go to in the browser to reach an HTTP app through
+Teleport, and it is the same for TCP app access.
+The domain name is expected to resolve to the public address of the Teleport Proxy.
+The app at this address is protected by TLS terminated at the Teleport Proxy,
+which usually has a publicly trusted wildcard TLS certificate for
+`*.teleport.example.com`.
+
+Teleport VNet will avoid overriding or conflicting with this domain name for a
+few reasons:
+
+1. To avoid incidentally breaking the currently supported flow if something goes
+   wrong with the VNet or it is just not running locally.
+1. While the tunnels VNet creates in the background will be authenticated by
+   mTLS, local clients will see it as an unauthenticated TCP endpoint.
+   When users connect to this domain name in the browser, we still want them to
+   see that it is protected by TLS.
+1. HTTP app domain names are sharable and they can be opened by anyone with an
+   account in the same cluster, regardless if they are running VNet or not.
+
+To break the conflict we can add a TLD/suffix to the domain name used in VNet.
+ICANN has recently assigned the `.internal` TLD for internal or private use
+applications so it seems like the appropriate choice.
+https://itp.cdn.icann.org/en/files/root-system/identification-tld-private-use-24-01-2024-en.pdf
+
+### Custom DNS zones
+
+Some users may want to use custom domain names for their apps in VNet.
+Maybe they currently access an app on their VPN at `api.example.private` and
+they want to migrate off their VPN to Teleport without changing all their
+software and scripts that currently reference that name.
+
+To do this they can set the `spec.vnet_addr` field in the app resource to
+`api.example.private`, but VNet will also need to know to configure itself as
+the DNS server for `*.example.private`, and what to do with other DNS requests
+matching `*.example.private` with no matching Teleport app.
+
+To accomodate this the Teleport admin can configure a custom DNS zone
+cluster-wide with a new `vnet` Teleport resource:
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  custom_dns_zones:
+  - suffix: .example.private
+    upstream_nameservers:
+    - 1.2.3.4
+    - 4.5.6.7
+  - suffix: .test.private
+```
+
+In this example, whenever the user is logged in to the cluster where this `vnet`
+configuration resource is present, VNet will install itself as the DNS
+nameserver for `*.example.private` and `*.test.private`.
+
+### DNS queries and IP address assignment
+
+This design tries to avoid making the VNet client maintain an up-to-date list of
+all apps in all clusters with assigned IPs.
+To that end, IPs will be dynamically assigned to apps the first time DNS is
+queried for that app.
+
+The VNet process will regulary poll for the currently logged-in root and leaf
+clusters to discover their proxy addresses and
+[custom DNS zones](#custom-dns-zones).
```

Comment:
Regular polling means that if you log in to a new root cluster and it doesn't have a resolver set up, we need to ask for admin privileges to set it up.

In Connect, we could signal that VNet is in a degraded state (by changing the status indicator from a green circle to an orange triangle for example). If the user tries to perform any action that makes use of VNet, we could say that VNet must be set up for cluster foo.example.com. They could also visit the UI I described in one of my previous comments and manually toggle VNet on for that new cluster. At this point they'd be asked for a password.

In tsh, the only two options we're left with is always limiting VNet only to clusters that were active at the start of the command (and clearly indicate that in stdout) or poll for changes and then prompt the user once we detect a new cluster with no resolver. Out of the two, it feels like just limiting VNet to active clusters is easier to understand as a user.

Reply:
Let's consider that the most common case is that the user will only ever log in to one cluster. UX will be good here. If DNS is already set up we can just leave it and don't need root again on relogins (but we will need it at least once every time Connect starts to create the TUN and configure the IP routes). This does raise the question of whether we should be removing the DNS configuration every time VNet or connect stops and if there is a good way to do this - we can try to handle the exit signal and clean up

Perhaps we don't need to poll - we can have a closer integration with Connect, every time you log in to a new cluster in Connect we can do the admin prompt and handle the DNS setup

Reply:
Oh there is also the idea that we could just leave the admin process running in the background permanently and communicate with it whenever we need to change some configuration

Reply:
I removed the polling from the doc. I think connect should signal to vnet when logging in to a new cluster and we can initiate the admin process immediately

Reply:
Initially, I understood your suggestion as "ask for prompt on each login, even if VNet is not running". But now that I think about it, it wouldn't make much sense. But yeah, we should be able to ask for root on new login if VNet is already running.

As you said, this will provide a good UX for most of users and slightly degrade it for people who regularly use more than one cluster. [Edoardo pulled the number of clusters per user in Connect](https://reporting-grafana-prod.platform.teleport.sh/d/c770d134-9a11-4ae3-a593-57d4a4eee5c3/teleport-connect?orgId=1) and as we expected, the overwhelming majority uses a single cluster. However, another thing that's true is that at the moment [people use Connect mostly for SSH and db access](https://reporting-onprem.platform.teleport.sh/insights/eU1uMfDC). I wouldn't want those users to see the root prompt since they probably won't use VNet at first.

Once we move off osascript it won't matter as we should be able to use privileged APIs without any prompts.